init/nitro/main.c

```diff
@@ -0,0 +1,231 @@
+// SPDX-License-Identifier: Apache-2.0
+
+#define _GNU_SOURCE
+#include <errno.h>
+#include <fcntl.h>
+#include <signal.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <unistd.h>
+#include <sys/mount.h>
+#include <sys/reboot.h>
+#include <sys/socket.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <sys/syscall.h>
+#include <linux/vm_sockets.h>
+
+#define finit_module(fd, param_values, flags) (int)syscall(__NR_finit_module, fd, param_values, flags)
+
+#define HEART_BEAT 0xb7
+#define VSOCK_CID 3
+#define VSOCK_PORT 9000
+
+/*
+ * Block or unblock signals.
+ *
+ * NOTE: All signals are blocked before the devies or console are set up.
+ *       Therefore, perror output may not be displayed if a failure occurs
+ *       during this setup.
+ */
+int
+sig_mask(int mask)
+{
+    sigset_t set;
+    int ret;
+
+    // Initialize the signal set to the complete set of supported signals.
+    ret = sigfillset(&set);
+    if (ret < 0) {
+        perror("sigfillset");
+        return ret;
+    }
+
+    // Block/unblock the signals. This essentially blocks/unblocks all signals
+    // to the process.
+    ret = sigprocmask(mask, &set, 0);
+    if (ret < 0) {
+        perror("sigprocmask");
+        return ret;
+    }
+
+    return 0;
+}
+
+/*
+ * Initialize a devtmpfs at /dev.
+ */
+int
+dev_init()
+{
+    int ret;
+
+    ret = mount("dev", "/dev", "devtmpfs", MS_NOSUID | MS_NOEXEC, NULL);
+    if (ret < 0) {
+        perror("mount");
```

Comment:
IMO it would be beneficial to have more descriptive error messages; maybe something along the lines of "failed to mount /dev as devtmpfs" for this specific instance.